src/router/filter.rs

```diff
@@ -0,0 +1,149 @@
+use bytes::{Buf, BufMut};
+use helium_proto::Eui;
+use std::{fmt, sync::Arc};
+use xorf::{Filter as XorFilter, Xor16};
+use xxhash_rust::xxh64::Xxh64;
+
+#[derive(Clone)]
+pub struct EuiFilter(Arc<Xor16>);
+#[derive(Clone, Debug)]
+pub struct DevAddrFilter {
+    base: u32,
+    size: u32,
+}
+
+impl fmt::Debug for EuiFilter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("EuiFilter")
+            .field("seed", &self.0.seed)
+            .field("blocks", &self.0.block_length)
+            .field("fingerprints", &self.0.len())
+            .finish()
+    }
+}
+
+impl EuiFilter {
+    pub fn from_bin<D: AsRef<[u8]>>(data: D) -> Self {
+        let mut buf = data.as_ref();
+        let seed = buf.get_u64_le();
+        let block_length = buf.get_u64_le() as usize;
+        let mut filters: Vec<u16> = Vec::with_capacity(block_length * 3);
+        for _ in 0..block_length * 3 {
+            filters.push(buf.get_u16_le());
+        }
+        Self(Arc::new(Xor16 {
+            seed,
+            block_length,
+            fingerprints: filters.into_boxed_slice(),
+        }))
+    }
+
+    pub fn contains(&self, eui: &Eui) -> bool {
+        let Eui { deveui, appeui } = eui;
+        let mut data = [0u8; 16];
+        let mut buf = &mut data[..];
+        buf.put_u64_le(*deveui);
+        buf.put_u64_le(*appeui);
+        let mut hasher = Xxh64::new(0);
+        hasher.update(&data);
+        let hash = hasher.digest();
+        self.0.contains(&hash)
+    }
+}
+
+const BITS_23: u64 = 8388607; // biggest unsigned number in 23 bits
+const BITS_25: u64 = 33554431; // biggest unsigned number in 25 bits
```

Comment:
Could derive or use binary literal instead of a decimal:
```rust
const BITS_25: u64 = 2_u64.pow(25) - 1; // biggest unsigned number in 25 bits
```
```rust
const BITS_25: u64 = 0b0001_1111_1111_1111_1111_1111_1111; // biggest unsigned number in 25 bits
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=16cfbc866b3b73e70420a1557944a7dd